### PR TITLE
Introducing the golang plugin [v3]

### PR DESCRIPTION
**Pull request by apahim**
v3:
 - Remove duplicated text from docs.
- Move the `go` binary probe up.
- Fix the `go` subtest filter parser.
- Accept files and directories as test references.
- Accept `.` as test reference.
- Fix the `GOROOT` and `GOPATH` handler (they are not paths, they are 'lists of paths').

v2: #2250 
- Try to find packages in also in $GOROOT (in addition to $GOPATH)
- Fix issue with non-existing paths.
- Don't discover any test when no reference is provided.

v1: #2245

### CODE REVIEW

File: optional_plugins/golang/avocado_golang/__init__.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        if ':' in url:
+            url, _tests_filter = url.split(':', 1)
+            parsed_filter = re.split(r'(?<!\\)/', _tests_filter, 1)
+            _tests_filter = parsed_filter[0]
+            if len(parsed_filter) > 1:
+                subtest = parsed_filter[1]
+            tests_filter = re.compile(_tests_filter)
+
+        # When a file is provided
+        if os.path.isfile(url):
```
Comment (ldoktor) [inline optional_plugins/golang/avocado_golang/__init__.py]: Isn't this treatment unnecessary complex? I'd expect it to work the same way as `GOROOT` so plain `package_paths.append('.')` would IMO work. Anyway I don't know that for sure and this (even so it's long) should work as well. We can modify it based on GO users requests later.

File: optional_plugins/golang/avocado_golang/__init__.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                    continue
+                avocado_suite.append((GolangTest, {'name': test_name,
+                                                   'subtest': subtest}))
+
+            return avocado_suite or self._no_tests(which_tests, url)
+
+        # When a directory is provided
+        if os.path.isdir(url):
+            files = self._find_files(url, recursive=False)
+            for file in files:
```
Comment (ldoktor) [inline optional_plugins/golang/avocado_golang/__init__.py]: `file` is a reserved builtin, please rename the variable... (path/paths is the usual choice)

Comment (apahim) [reply optional_plugins/golang/avocado_golang/__init__.py]: <o> sure sure

File: optional_plugins/golang/avocado_golang/__init__.py (hunk 1 above)
Comment (apahim) [reply optional_plugins/golang/avocado_golang/__init__.py]: As far as I could understand the `go test` behavior, there are some differences when you use a package reference when compared to file path references. Depending on the test and on $PWD, modules might not be loaded properly. That's why I think we should have a separate treatment for packages. If you don't mind, let's start with this and see where it leads us.

Comment (ldoktor) [reply optional_plugins/golang/avocado_golang/__init__.py]: yep, no problem, it just looked unnecessary complex but I only experienced with it briefly... We'll hear about that form go devels anyway 😄

**Pull request closed without merging**